File: src/main/kotlin/graph/크루스칼_알고리즘.java
package graph;

import java.util.*;

public class 크루스칼_알고리즘 {

    private static int[] parent;
    private static List<Edge> edges = new ArrayList<>();

    public static int findParent(int x){
        if(parent[x] == x) return x;
        return parent[x] = findParent(parent[x]);
    }

    public static void unionParent(int a, int b){
        a = findParent(a);
        b = findParent(b);
        if(a < b) parent[b] = a;
        else parent[a] = b;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int v = sc.nextInt();   //노드의 개수 입력받기
        int e = sc.nextInt();   //간선의 개수 입력받기

        //부모 테이블 초기화
        parent = new int[v+1];
        for(int i = 1; i <= v; i++){
            parent[i] = i;
        }

        //간선 정보 입력받기
        for(int i = 0; i < e; i++){
            int a = sc.nextInt();
            int b = sc.nextInt();
            int cost = sc.nextInt();

            edges.add(new Edge(a, b, cost));
        }
        //비용이 적은 순서대로 정렬
        Collections.sort(edges);

        //집합으로 묶기 수행
        int totalCost = 0;
        for(Edge edge: edges){
            int nodeA = edge.nodeA;
            int nodeB = edge.nodeB;
            int cost = edge.cost;
            //사이클이 발생하지 않았다면
            if(findParent(nodeA) != findParent(nodeB)){
                unionParent(nodeA, nodeB);
                totalCost += cost;
            }
        }

        System.out.println(totalCost);
    }

    static class Edge implements Comparable<Edge> {

        private int nodeA;
        private int nodeB;
        private int cost;

        public Edge(int nodeA, int nodeB, int cost){
            this.nodeA = nodeA;
            this.nodeB = nodeB;
            this.cost = cost;
        }

        public int getNodeA(){
            return this.nodeA;
        }

        public int getNodeB(){
            return this.nodeB;
        }

        public int getCost(){
            return this.cost;
        }

        @Override
        public int compareTo(Edge other){
            return this.cost - other.cost;
        }
    }
}

File: src/main/kotlin/graph/크루스칼_알고리즘.kt
package graph

import java.util.*
import kotlin.math.*


fun main(){
    val sc = Scanner(System.`in`)
    val v = sc.nextInt()
    val e = sc.nextInt()

    val parent = IntArray(v + 1)
    for(i in 1..v){
        parent[i] = i
    }

    fun findParent(x: Int): Int {
        if(x == parent[x]) return x
        return findParent(parent[x]).also { parent[x] = it }
    }

    fun unionParent(a: Int, b: Int){
        val parentA = findParent(a)
        val parentB = findParent(b)
        if(parentA < parentB) parent[parentB] = parent[parentA] else parent[parentA] = parent[parentB]
    }

    //간선에 대한 정보 입력받기
    val edges = arrayListOf<Edge>()
    for(i in 0 until e){
        val a = sc.nextInt()
        val b = sc.nextInt()
        val cost = sc.nextInt()
        edges.add(Edge(a, b, cost))
    }
    //거리 순서대로 정렬
    edges.sortBy { it.distance }

    //간선을 확인하면서 산장 트리를 만족하는 최소비용 측정
    var totalCost = 0
    for(edge in edges){
        val nodeA = edge.nodeA
        val nodeB = edge.nodeB
        val cost = edge.distance
        //사이클이 발생하지 않은 경우만 포함
        if(findParent(nodeA) != findParent(nodeB)){
            unionParent(nodeA, nodeB)
            totalCost += cost
        }
    }

    println(totalCost)
}

data class Edge(val nodeA: Int, val nodeB: Int, val distance: Int)